File: ext/types/executor_id.cpp
#include <iostream>
#include "types.h"

using namespace mesos;

ExecutorIDPtr toExecutorID(char* bs, int len)
{
  ExecutorIDPtr val = new ExecutorID();
  val->set_value(bs, len);
  return val;
}

int fromExecutorID(ExecutorIDPtr p, char** poke)
{
  *poke = (char*) p->mutable_value()->data();
  return p->mutable_value()->size();
}

void destroyExecutorID(ExecutorIDPtr p)
{
  delete p;
}

File: src/System/Mesos/Raw/ExecutorId.hs
module System.Mesos.Raw.ExecutorId where
import           System.Mesos.Internal

type ExecutorIDPtr = Ptr ExecutorID

foreign import ccall unsafe "ext/types.h toExecutorID" c_toExecutorID :: ToID ExecutorIDPtr

foreign import ccall unsafe "ext/types.h fromExecutorID" c_fromExecutorID :: FromID ExecutorIDPtr

foreign import ccall unsafe "ext/types.h destroyExecutorID" c_destroyExecutorID :: ExecutorIDPtr -> IO ()

instance CPPValue ExecutorID where
  marshal x = do
    (strp, l) <- cstring $ fromExecutorID x
    liftIO $ c_toExecutorID strp $ fromIntegral l

  unmarshal p = fmap ExecutorID $ do
    ptrPtr <- alloc
    len <- liftIO $ c_fromExecutorID p ptrPtr
    peekCString' (ptrPtr, len)

  destroy = c_destroyExecutorID
